[Grouping&Mapping] Properties Refactor

Guillar1 opened the pull request.

MindaugasButkus commented on common/config/rush/pnpm-lock.yaml:
```diff
@@ -270,9 +273,9 @@ importers:
+      '@itwin/imodel-browser-react': ^0.12.3
       '@itwin/imodel-components-react': ^3.2.0
-      '@itwin/insights-client': ^0.3.0
+      '@itwin/insights-client': 0.2.1
```
Don't downgrade the insights-client.

Guillar1 replied:
Forgot to run rush update after merge. Done.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/package.json:
```diff
@@ -53,13 +53,18 @@
-    "@itwin/imodel-browser-react": "^0.12.1",
-    "@itwin/insights-client": "^0.3.0",
+    "@itwin/imodel-browser-react": "^0.12.3",
+    "@itwin/insights-client": "0.3.0",
```
Don't drop the caret.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/SelectProject.tsx:
```diff
@@ -65,6 +65,7 @@ const SelectProject = ({ onSelect, onCancel }: SelectProjectProps) => {
   }, [getAccessToken]);
 
   const startSearch = useCallback(() => {
+    if(!searchInput) return;
```
This change looks unrelated.

Guillar1 replied:
It is unrelated. Small one line bug that stops a search from being made when no input is given.

MindaugasButkus replied:
Formatting inconsistency - there should be a space between `if` and `(`.

Guillar1 replied:
done

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/HorizontalTile.tsx:
```diff
@@ -6,23 +6,30 @@ import type { ReactNode } from "react";
 }
 
-export const HorizontalTile = ({ title, subText, onClickTitle, titleTooltip, subtextToolTip, actionGroup}: HorizontalTileProps) => {
+export const HorizontalTile = ({ title, subText, onClick, onClickTitle, titleTooltip, subtextToolTip, actionGroup, selected, dragHandle}: HorizontalTileProps) => {
```
I don't think destructuring helps when you have that many properties. I think simply using `props.onClick` would be good. Now the props do not fit into the screen.

Guillar1 replied:
done

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyTable.tsx:
```diff
@@ -82,13 +81,6 @@ const GroupPropertyTable = ({
                   menuItems={(close: () => void) => [
-                    <MenuItem
-                      key={0}
-                      onClick={() => onGroupPropertyModify(value)}
```
Is it no longer possible to modify the property? I think you said that you were able to bring this back.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.scss:
```diff
@@ -2,31 +2,99 @@
+  min-width: 100%;
+  height: 75vh;
+
+  .gutter {
```
Non-gmw selectors. Where do they come from? What is the background-image below?

Guillar1 replied:
All of these selectors were for the split library itself. I managed to change it.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/SortableHorizontalTile.tsx:
```diff
@@ -0,0 +1,55 @@
+    if (!isDragging) {
+      return;
+    }
+    document.body.style.cursor = "grabbing";
```
This looks like a hack. What does this do?

Guillar1 replied:
Globally change the cursor to grabbing when actively dragging around the tile.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/SortableHorizontalTile.tsx:
```diff
@@ -0,0 +1,55 @@
+import React, { useEffect } from "react";
```
Missing copyright header.

Guillar1 replied:
Huh, the git hook didn't run the script to insert it. Manually added it.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.scss:
```diff
@@ -2,31 +2,99 @@
+.gmw-properties-list {
+  display: flex;
+  flex-direction: column;
+  gap: 7.5px;
```
There should be a variable that will match what you need.

Guillar1 replied:
Not for vertical spacing, $iui-baseline is typically used. All the other spacing variables are meant to be used in horizontal spacing.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:
```diff
@@ -68,98 +91,138 @@ export const quantityTypesSelectionOptions: SelectOption<QuantityType>[] = [
+    case "number":
+      return DataType.Number;
+    default:
+      return DataType.Undefined;
```
I think it would be better to default to `DataType.String`. Sending `DataType.Undefined` to Reporting API will cause a BadRequest.

Guillar1 replied:
Done.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:
```diff
@@ -204,55 +310,135 @@ const extractProperties = (
+      return [
+        {
+          ...ecProperty,
+          ecPropertyName: [...property.propertyTraversal.slice(0, -1), "Material", "UserLabel"].join("."),
```
Why the `slice` for Material?

Guillar1 replied:
The property name that Presentation returns is PhysicalMaterial. I slice it out and replace it with Material because that is the actual column name.

MindaugasButkus replied:
This looks like a very weird and unnecessary edge case. See the BisCore schema: https://github.com/iTwin/bis-schemas/blob/master/Domains/0-Core/BisCore.ecschema.xml

PhysicalMaterial everywhere. How can it be the "actual column name"?

Guillar1 replied:
Good catch. I'm not sure how I arrived to "Material". Must have confused another schema when I looked into this a while back.

Guillar1 replied:
done

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:
```diff
@@ -204,55 +310,135 @@ const extractProperties = (
+      return [
+        {
+          ...ecProperty,
+          ecPropertyName: [
```
Keep the formatting consistent. Here you add each array element to a new line. Below you keep everything on a single line. Pick one.

Guillar1 replied:
Done

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:
```diff
@@ -68,98 +91,138 @@ export const quantityTypesSelectionOptions: SelectOption<QuantityType>[] = [
+  //  It belongs to this parent class
+  const parentPropertyClassName = propertyField.parent?.contentClassInfo.name;
+  const primitiveNavigationClass = propertyField.properties[0].property.navigationPropertyInfo?.classInfo.name ?? "";
+  // Presentation treats primitive navigation classes as longs. Handling this special case.
```
What do you mean by this comment? Are these longs the ECInstanceId of the ECClass (in meta.ECClassDef)?

Guillar1 replied:
Yes, it is the ECInstanceId of the ECClass stored as a long in memory in the C++ layer of Presentation when dealing with Navigations. Not sure of the intention is. I can link the line this happens if requested due to it being in Azure DevOps.

MindaugasButkus replied:
Please be a bit more specific what that long is. If it is the value of `ECInstanceId` of the class in the `meta.ECClassDef` table, then say so in the comment.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:
```diff
@@ -68,98 +91,138 @@ export const quantityTypesSelectionOptions: SelectOption<QuantityType>[] = [
+  const primitiveNavigationClass = propertyField.properties[0].property.navigationPropertyInfo?.classInfo.name ?? "";
+  // Presentation treats primitive navigation classes as longs. Handling this special case.
+  const type = primitiveNavigationClass ? DataType.String : convertType(propertyField.properties[0].property.type);
+  const actualECClassName = propertyField.properties[0].property.classInfo.name;
```
In what cases will the `propertyField.properties` array contain more than 1 entry?

Guillar1 replied:
I believe it happens during property merging. Forgot to consider this case. https://github.com/iTwin/itwinjs-core/blob/master/docs/presentation/content/Terminology.md#property-merging
I wrote the logic to deal with this but I cannot find an iModel where this edge case occurs.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:
```diff
@@ -22,31 +22,54 @@ import {
+import { HorizontalTile } from "./HorizontalTile";
+import { DataType, QuantityType } from "@itwin/insights-client";
+import type { ECProperty, GroupProperty, GroupPropertyCreate } from "@itwin/insights-client";
+import { SvgClose, SvgDragHandleVertical, SvgRemove, SvgSearch } from "@itwin/itwinui-icons-react";
```
I get this error: `Module '"@itwin/itwinui-icons-react"' has no exported member 'SvgDragHandleVertical'.`

Guillar1 replied:
Should be fixed.

Guillar1 replied:
Should be fixed now really.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/package.json:
```diff
@@ -102,6 +107,7 @@
     "react-table": "^7.7.0",
     "redux": "^4.1.0",
     "rimraf": "^3.0.2",
+    "split.js": "^1.6.0",
```
Why is split.js a devDependency? I see that react-split has a dependency on split.js, but I don't understand why we need to explicitly add it into our package.json.

Guillar1 replied:
This wasn't suppose to be here. I used it to hoist up the package to check the types when developing this. I'll let you know when this is review ready. I am taking my laptop to another location and I pushed what I had just incase something happens to it.

Guillar1 replied:
Removed

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:
```diff
@@ -68,98 +92,141 @@ export const quantityTypesSelectionOptions: SelectOption<QuantityType>[] = [
+  propertyField: PropertiesField
+): PropertyMetaData[] => {
+  const properties = new Array<PropertyMetaData>();
+  for(const property of propertyField.properties){
```
Rewrite this `for` with a `map`.

Guillar1 replied:
Done. And replaced other similar situations with `flatMap`.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:
```diff
@@ -3,6 +3,7 @@
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
 import type { IModelConnection } from "@itwin/core-frontend";
```
The length of this file is ridiculous. Move out all the property retrieval / construction logic out so that only the UI part is left in this file.

Guillar1 replied:
Moved out all property constructed relation functions to its own file.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:
```diff
@@ -264,90 +97,83 @@ const GroupPropertyAction = ({
-    useState<Map<string, Property[]>>();
-  const [ecProperties, setEcProperties] = useState<ECProperty[]>(
+  const [selectedProperties, setSelectedProperties] = useState<PropertyMetaData[]>([]);
+  const [propertiesMetaData, setPropertiesMetaData] = useState<PropertyMetaData[]>(
```
Why is this formatted differently than the line above? They are the same length.

Guillar1 replied:
One line again.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:
```diff
@@ -264,90 +97,83 @@ const GroupPropertyAction = ({
+  const filteredProperties = useMemo(
+    () =>
+      propertiesMetaData.filter((p) =>
+        [p.label, p.categoryLabel, p.actualECClassName]
```
There is a chance that this might produce unexpected results. I think this would be better:
```
[p.label, p.categoryLabel, p.actualECClassName]
  .map((l) => l.toLowerCase())
  .some((l) => l.includes(activeSearchInput.toLowerCase()))
```

Guillar1 replied:
Oh right, forgot to lowercase.

Guillar1 replied:
I like this so much better. Thanks :).

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts:
```diff
@@ -0,0 +1,360 @@
+  let notFound = false;
+  while (ecPropertiesCopy.length !== 0) {
+    for (let i = 0; i < propertiesMetaData.length; i++) {
+      const generatedProperty = convertToECProperties(propertiesMetaData[i]);
```
Convert all `propertiesMetaData` with a `map` function once instead of doing the conversion for all of them for each iteration of the outer `while` loop.

Guillar1 replied:
I will come back to this later as the currently implementation works.

Guillar1 commented:
@MindaugasButkus FYI, I changed where the category was being displayed. It was getting really cramped with long category names so I moved it to the subText of the tile. 
![image](https://user-images.githubusercontent.com/26608199/192182773-ad310db9-cca2-4962-a79f-8215ca055acf.png)

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts:
```diff
@@ -0,0 +1,360 @@
+import type { IModelConnection } from "@itwin/core-frontend";
```
Missing copyright header.

Guillar1 replied:
Done.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts:
```diff
@@ -0,0 +1,360 @@
+
+export interface PropertyMetaData {
+  // Display label
+  label: string;
```
Instead of adding comments, you could just rename the properties:
```
label -> displayLabel
schema -> sourceSchema
className -> sourceClassName
type -> propertyType
```
etc.

Guillar1 replied:
Changed them. Let me know what you think.

MindaugasButkus replied:
Looks better now.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts:
```diff
@@ -0,0 +1,364 @@
+          // Navigation properties
+          case RelationshipMeaning.RelatedInstance: {
+            if (
+              // Deal with a TypeDefinition
```
Why is there a special case for TypeDefinition?

Guillar1 replied:
I have not seen another type of a related instance that weren't the primitive navigations.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts:
```diff
@@ -0,0 +1,364 @@
+                  .name === "BisCore:ElementOwnsMultiAspects")
+            ) {
+              const fullClassName = nestedContentField.contentClassInfo.name;
+              const schema = fullClassName.split(":")[0];
```
Is the `schema` and `class` in `fullClassName` always separated by `:`, or can it be also separated by `.`?

Guillar1 replied:
Always `:`

MindaugasButkus commented on common/config/rush/pnpm-lock.yaml:
```diff
@@ -668,24 +678,24 @@ importers:
-      '@itwin/imodel-components-react': 3.2.4_284f744975b7010765e33cd82cf5abd5
-      '@itwin/presentation-common': 3.2.4_292635cddc94cb9e89eee37a7ec1b018
-      '@itwin/presentation-frontend': 3.2.4_17579ccf48e56cfd5df4551deb49a966
+      '@itwin/core-react': 3.1.2_51d9f14b6be2a2a40dc41eb8b80a5b52
```
Why the downgrades?

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:
```diff
@@ -264,90 +97,80 @@ const GroupPropertyAction = ({
+  const [searchInput, setSearchInput] = useState<string>("");
+  const [activeSearchInput, setActiveSearchInput] = useState<string>("");
+  const [searched, setSearched] = useState<boolean>(false);
+  const [activeDragProperty, setActiveDragProperty] = useState<PropertyMetaData>();
```
No default provided for the state which cannot be `undefined`. Change the type or provide a default.

Guillar1 replied:
Changed the type. I figured that not passing anything was implicit enough.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:
```diff
@@ -264,90 +97,80 @@ const GroupPropertyAction = ({
-      const content = await Presentation.presentation.getContentDescriptor(
-        requestOptions
-      );
+  const handleDragStart = (event: DragStartEvent) => {
```
This might need to be a callback, because it depends on `selectedProperties` value.

Guillar1 replied:
Both needed callbacks actually, the `onHandle` props most likely have these functions as dependencies as well.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyAction.tsx:
```diff
@@ -451,76 +239,46 @@ const GroupPropertyAction = ({
-        <Fieldset className='gmw-property-options' legend='Property Details'>
-          <Small className='gmw-field-legend'>
+        <Fieldset disabled={isLoading} className='gmw-property-options' legend='Property Details'>
+          <Small className='field-legend'>
```
Why did you remove the `gmw-` prefix?

Guillar1 replied:
Bad merge brought it back.

MindaugasButkus commented on packages/itwin/grouping-mapping-widget/src/widget/components/HorizontalTile.tsx:
```diff
@@ -6,26 +6,33 @@ import type { ReactNode } from "react";
+      <div className="gmw-body-container">
+        {props.dragHandle}
+        <div className="gmw-body">
+          <Text className={classNames("gmw-body-text", {"iui-anchor": props.onClickTitle})} onClick={props.onClickTitle} variant="body" title={props.titleTooltip}>{props.title}</Text>
```
I think it might be better to `!!props.onClickTitle` when setting to the `iui-anchor` property. That way it will get converted to boolean.

Guillar1 replied:
Sure done.
